### PR TITLE
Perform short pre-optimisation for IRR

### PR DESCRIPTION
**Pull request by pascalhein**
When computing IRR, the NPV curve may be very degenerate (such as in #1904):
curve
In this example, the Newton iteration only converges if the initial guess x0 is to the left of the extreme point (which is at approx. x = -0.9975), but currently, the initial guess is -0.95. (Note that the code actually computes "IRR+1", hence the initial guess x0 = 0.05.)

This commit attempts to find an initial guess in the interval (-1,0) first, via interval halving.

### CODE REVIEW

File: name.abuchen.portfolio/src/name/abuchen/portfolio/math/IRR.java [hunk 1]
```diff
@@ -11,10 +12,39 @@ private IRR()
...
+
     public static double calculate(List<LocalDate> dates, List<Double> values)
     {
         Function npv = new NPVFunction(dates, values);
         Function derivative = new PseudoDerivativeFunction(npv);
-        return NewtonGoalSeek.seek(npv, derivative, 0.05d) - 1;
+
+        // find a crude initial guess in interval (0,1)
+        // npv(0) is undefined, but diverges with sign given by last term
+        double fLeft = Double.POSITIVE_INFINITY * values.get(values.size() - 1);
```
Comment (chirlu) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/math/IRR.java:42]: For better readability, you could consider using `Math.copySign()`.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/math/IRR.java [hunk 2]
```diff
@@ -11,10 +12,39 @@ private IRR()
...
+        // npv(1) is the sum of undiscounted flows
+        double fRight = values.stream().collect(Collectors.summingDouble(f -> f));
+        // if they have the same sign, we hopefully don't have a very extreme case, so just guess 5%
+        double guess = fLeft * fRight < 0 ? halving(npv, 0, 1, fLeft, fRight) : 1.05;
```
Comment (chirlu) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/math/IRR.java:46]: In a similar way, using `Math.signum()` might make the condition more readable.

**Review by chirlu — APPROVED** [review]

**Comment by chirlu** [comment]
@buchen: I’d like to remind you of this PR. Any chance it can get in?

**Comment by CK1-CK** [comment]
Unfortunatly the Bug is still present. 

see https://github.com/buchen/portfolio/issues/1904#issuecomment-885721821

**Pull request merged**